tests/core/pipeline/test_pipeline_manager_with_sql_repo.py

```diff
@@ -0,0 +1,525 @@
+    # only create intermediate data node once
+    Config.configure_job_executions(mode=JobConfig._DEVELOPMENT_MODE)
+    Config.global_config.repository_type = "sql"
+
+    _PipelineManager._repository = _PipelineRepositoryFactory._build_repository()
+    _SchedulerFactory._build_dispatcher()
+
+    dn_config_1 = Config.configure_data_node("foo", "in_memory", Scope.PIPELINE, default_data=1)
+    dn_config_2 = Config.configure_data_node("bar", "in_memory", Scope.PIPELINE, default_data=0)
+    dn_config_6 = Config.configure_data_node("baz", "in_memory", Scope.PIPELINE, default_data=0)
+
+    task_config_mult_by_two = Config.configure_task("mult_by_two", mult_by_two, [dn_config_1], dn_config_2)
+    task_config_mult_by_3 = Config.configure_task("mult_by_3", mult_by_3, [dn_config_2], dn_config_6)
+    pipeline_config = Config.configure_pipeline("by_6", [task_config_mult_by_two, task_config_mult_by_3])
+    # dn_1 ---> mult_by_two ---> dn_2 ---> mult_by_3 ---> dn_6
+
+    assert len(_DataManager._get_all()) == 0
+    assert len(_TaskManager._get_all()) == 0
+
+    pipeline = _PipelineManager._get_or_create(pipeline_config)
+
+    assert len(_DataManager._get_all()) == 3
+    assert len(_TaskManager._get_all()) == 2
+    assert len(pipeline._get_sorted_tasks()) == 2
+    assert pipeline.foo.read() == 1
+    assert pipeline.bar.read() == 0
+    assert pipeline.baz.read() == 0
+    assert pipeline._get_sorted_tasks()[0][0].config_id == task_config_mult_by_two.id
+    assert pipeline._get_sorted_tasks()[1][0].config_id == task_config_mult_by_3.id
+
+    _PipelineManager._submit(pipeline.id)
+    assert pipeline.foo.read() == 1
+    assert pipeline.bar.read() == 2
+    assert pipeline.baz.read() == 6
+
+    pipeline.foo.write("new data value")
+    assert pipeline.foo.read() == "new data value"
+    assert pipeline.bar.read() == 2
+    assert pipeline.baz.read() == 6
+
+    pipeline.bar.write(7)
+    assert pipeline.foo.read() == "new data value"
+    assert pipeline.bar.read() == 7
+    assert pipeline.baz.read() == 6
+
+    with pytest.raises(AttributeError):
+        pipeline.WRONG.write(7)
+
+
+def test_do_not_recreate_existing_pipeline_except_same_config():
+    Config.global_config.repository_type = "sql"
+    _PipelineManager._repository = _PipelineRepositoryFactory._build_repository()
+    dn_input_config_scope_scenario = Config.configure_data_node("my_input", "in_memory", scope=Scope.SCENARIO)
+    dn_output_config_scope_scenario = Config.configure_data_node("my_output", "in_memory", scope=Scope.SCENARIO)
+    task_config = Config.configure_task(
+        "task_config", print, dn_input_config_scope_scenario, dn_output_config_scope_scenario
+    )
+    pipeline_config = Config.configure_pipeline("pipeline_config_1", [task_config])
+
+    # Scope is scenario
+    pipeline_1 = _PipelineManager._get_or_create(pipeline_config)
+    assert len(_PipelineManager._get_all()) == 1
+    # TODO: failed because couldn't detect the already created pipeline
```

Comment:
This should have been fixed already, I believe.

Reply:
agree! Please leave your comment here for now so I can remember to remove the line later :D